### PR TITLE
fix(ui5-shellbar): fix error when there are too many resizes

### PR DESCRIPTION
**Pull request by IlianaB**
Debounce the handleResize function to be executed just once in a given
interval of time.

Fixes: #3312

### CODE REVIEW

File: packages/fiori/src/ShellBar.js [hunk 1]
```diff
@@ -464,14 +466,24 @@ class ShellBar extends UI5Element {
...
+				await this._getResponsivePopover();
+				this.overflowPopover.close();
+				this._overflowActions();
+			}.bind(this), HANDLE_RESIZE_DEBOUNCE_RATE);
 		};
 
 		this.i18nBundle = getI18nBundle("@ui5/webcomponents-fiori");
 	}
 
+	_debounce = (fn, delay) => {
```
Comment (nnaydenow) [inline packages/fiori/src/ShellBar.js:479]: lint is failing here

File: packages/fiori/src/ShellBar.js [hunk 2]
```diff
@@ -464,14 +466,24 @@ class ShellBar extends UI5Element {
 		};
 
 		this._handleResize = async event => {
-			await this._getResponsivePopover();
-			this.overflowPopover.close();
-			this._overflowActions();
+			this._debounce(async function() {
```
Comment (vladitasev) [inline packages/fiori/src/ShellBar.js:469]: Do not use a normal function, unless needed. Arrow function preferred:

`this._debounce(async () => {
......
}, HANDLE_RESIZE_)`

File: packages/fiori/src/ShellBar.js [hunk 3]
```diff
@@ -464,14 +466,24 @@ class ShellBar extends UI5Element {
...
+				await this._getResponsivePopover();
+				this.overflowPopover.close();
+				this._overflowActions();
+			}.bind(this), HANDLE_RESIZE_DEBOUNCE_RATE);
```
Comment (vladitasev) [inline packages/fiori/src/ShellBar.js:473]: `.bind(this)` becomes unnecessary then

File: packages/fiori/src/ShellBar.js (hunk 1 above)
Comment (vladitasev) [inline packages/fiori/src/ShellBar.js:479]: What you probably meant is:

`_debounce(fn, delay) {`

(class method)

**Review by vladitasev — COMMENTED** [review]
I'll think how we can make the change to the tests smaller. In the meantime, I've commented on some changes in the source file.

File: packages/fiori/src/ShellBar.js [hunk 4]
```diff
@@ -464,14 +466,24 @@ class ShellBar extends UI5Element {
...
+				await this._getResponsivePopover();
+				this.overflowPopover.close();
+				this._overflowActions();
+			}, HANDLE_RESIZE_DEBOUNCE_RATE);
 		};
 
 		this.i18nBundle = getI18nBundle("@ui5/webcomponents-fiori");
 	}
 
+	_debounce(fn, delay) {
```
Comment (fifoosid) [inline packages/fiori/src/ShellBar.js:479]: There is a debounce function in the base project which should be imported and used.
It is located here: ```packages/base/src/util/debounce.js```

**Review by fifoosid — CHANGES_REQUESTED** [review]

File: packages/fiori/src/ShellBar.js (hunk 4 above)
Comment (IlianaB) [reply packages/fiori/src/ShellBar.js:479]: Yes, but it needs to be changed. We have already discussed that with @ilhan007 

The util debounce function will always execute only the latest function which was passed in the given interval. Currently, there are three different controls, calling it, with the shellbar there will be four controls. If some/all of them (or even different instances of these controls) call debounce at the same time, only the last function will be executed, which is not OK. That's why here we have this._debounceInterval per each instance. 

In my opinion, general utility function should be even more flexible and implemented as part of a different task with clear requirements.

**Review by ilhan007 — APPROVED** [review]

**Comment by nnaydenow** [comment]
Close: #3022

**Pull request merged**